seccomp.go

```diff
@@ -540,6 +632,27 @@ func NewFilter(defaultAction ScmpAction) (*ScmpFilter, error) {
 	return filter, nil
 }
 
+// SetTsync sets or clears the filter's thread-sync (TSYNC) attribute. When set, this attribute
+// tells the kernel to synchronize all threads of the calling process to the same seccomp filter.
+// When using filters with the seccomp notification action (ActNotify), the TSYNC attribute
+// must be cleared prior to loading the filter. Refer to the seccomp manual page (seccomp(2)) for
+// further details.
+func (f *ScmpFilter) SetTsync(val bool) error {
```

Comment:
Perhaps I'm reading this too quickly, but if we always enable the `TSYNC` bit in `NewFilter()` and there is little reason to disable the `TSYNC` bit in golang, why does the `SetTsync()` function exist?

Reply:
When seccomp notification was initially implemented in Linux, it was initially implemented without support for `TSYNC`. See this commit that makes `NEW_LISTENER` and `TSYNC` flags exclusive, meaning the kernel rejects filters with both flags:
https://github.com/torvalds/linux/commit/7a0df7fbc14505e2e2be19ed08654a09e1ed5bf6

Then Linux 5.7 introduced a new API to enable seccomp notification with tsync together in this commit:
https://github.com/torvalds/linux/commit/51891498f2da78ee64dfad88fa53c9e85fb50abf

But for unfortunate users of Linux < 5.7, they need to disable `TSYNC` and use `runtime.LockOSThread()` instead to ensure that, upon forking, the new process retains the filter:
```
if seccompAPILevel == 5 {
  filter.SetTsync(false)
```

For users of Linux >= 5.7 with a sufficiently recent libseccomp (API Level >= 6), I agree that `SetTsync()` has little reason to exist. But I would like libseccomp-golang to work with older libseccomp and older kernels. And I think that the function `SetTsync()` unfortunately needs to be exposed to the users of the libseccomp-golang library so that they know they have to use `runtime.LockOSThread()` (it could cause incompatibility issues if the `TSYNC` flag was removed automatically by the library without explicit request).

Reply:
Ah ha, gotcha.  If you look at the main libseccomp library you'll see we struggled with this too.

With golang, I believe that if we have to choose between the `TSYNC` and notify functionality, `TSYNC` should win.  Instead of adding a new API and exposing something we really don't want to expose (disabling `TSYNC` for golang programs), I'd much rather restrict the notification functionality in the golang bindings to only those systems with an API level greater than or equal to six (supports the simultaneous use of `TSYNC` and notify).

Do you have any thoughts on this @drakenclimber?

Reply:
It sounds fine to me for what I want to do with runc.

For the history, we initially wrote [runc PR 2682](https://github.com/opencontainers/runc/pull/2682) using `SetTsync()` to be compatible with Linux < 5.7 (see [InitSeccomp()](https://github.com/opencontainers/runc/blob/1751ec3b27890e1e33ea3f23f61bd250b6b493b6/libcontainer/seccomp/seccomp_linux.go#L67-L82)) but we removed that in the last version of the runc PR (see [filterFlags()](https://github.com/opencontainers/runc/blob/c25bb2bc8556aee67824ac35a3224d12d4e50de6/libcontainer/seccomp/patchbpf/enosys_linux.go#L580-L586)) for keeping things simple, since we use Linux >= 5.7 anyway.

However, [nestybox/sysbox-runc](https://github.com/nestybox/sysbox-runc/) uses `SetTsync()` (see [prepNotify()](https://github.com/nestybox/sysbox-runc/blob/9d050e41cba7b32f43830b22c9a2bca97e01fa6f/libcontainer/seccomp/seccomp_linux.go#L311)) and might prefer to keep compatibility with older Linux distributions that use Linux < 5.7, which is possible to do in a safe way given how runc uses `runtime.LockOSThread()`.

cc @ctalledo

Reply:
Hi @alban, @pcmoore,

Given that Linux supports seccomp-notify since earlier than 5.7, my preference would be to have seccomp-golang expose `SetTsync()`. But I understand that it's something that should ideally not be exposed as part of the seccomp-golang API, and that starting with Linux 5.7 there is less need to expose it (though IIRC calling `runtime.LockOSThread()` would still be needed). 

As far as the usage of seccomp-golang in [nestybox/sysbox-runc](https://github.com/nestybox/sysbox-runc), we definitely need sysbox-runc to support kernels < 5.7, as most users are on such kernels. Thus, if the consensus is that `SetTsync()` be removed from the seccomp-golang API, we would unfortunately need to continue using a custom version of seccomp-golang rather than the upstream version, until the need to support kernels < 5.7 goes away. 

Thanks again for all good work in driving this PR to completion, much appreciated!

Reply:
> Ah ha, gotcha. If you look at the main libseccomp library you'll see we struggled with this too.
> 
> With golang, I believe that if we have to choose between the `TSYNC` and notify functionality, `TSYNC` should win. Instead of adding a new API and exposing something we really don't want to expose (disabling `TSYNC` for golang programs), I'd much rather restrict the notification functionality in the golang bindings to only those systems with an API level greater than or equal to six (supports the simultaneous use of `TSYNC` and notify).
> 
> Do you have any thoughts on this @drakenclimber?

Yes, this was real challenge in libseccomp.  Based on the comments above, I think we have to expose `SetTsync()` for legacy libseccomp-go users and users on older kernels.  But we should probably add a strong comment in it that warns of its dangers.

Reply:
Added a big warning for this, then!

Reply:
Sorry all, I still think adding `SetTsync()` to the public API is a mistake.  My concern is that under Go it is far too easy to misuse this API and end up messing up the seccomp filtering for the process.  Further, the `SetTsync()` API is really only needed to workaround some limitations/bugs in a limited number of kernel releases; it's not necessary using a current or future kernels.

As mentioned above, I believe the best move here is to not expose `SetTsync()` and limit the notification functionality to only those systems with an API level greater than or equal to six.  Exposing a problematic API only to enable new functionality doesn't seem like a smart thing to do as far as I'm concerned.

Reply:
@pcmoore no problem, I've updated to remove `SetTsync()` now. Same notes as mentioned [here][link] apply. PTAL :)

[link]: https://github.com/seccomp/libseccomp-golang/pull/59#issuecomment-817791278